design_docs/selfhealing_featuretoggle.md

```diff
@@ -0,0 +1,160 @@
+
+# Keptn Feature-Toggle Self-Healing
+
+This document describe how Keptn leverages feature toggle frameworks to enable self-healing for applications by automatically toggling feature toggles in response to problem tickets that are reported to Keptn.
+
+At first, we describe how to write remediation actions agnostic to the actual feature toggle framework used. Secondly, we describe how a user would work with Keptn as a control plane to toggle feature flags for automated operations of the users applications. 
+Please note that in the current version, Keptn will provide built-in functionality for the open-source feature toggle framework [Unleash](https://unleash.github.io/), with other feature toggle platforms to follow.
+
+
+# Prerequisites
+
+
+
+*   Running Keptn Installation
+*   Running [Unleash](https://unleash.github.io/) server as the feature toggle framework
+*   Application with feature flags connected to the Unleash server
+*   Remediation file with feature flags to toggle in case of issues
+*   Running monitoring solution for the application (right now Dynatrace is supported)
+
+
+# Remediation file
+
+The remediation file is necessary to define which remediation action should be taken in response to a problem ticket that is sent to Keptn. Keptn’s internal remediation service will then determine the actions to take and trigger them accordingly. 
+
+Here’s an example of a remediation file:
+
+remediation.yaml
+
+```yaml
+remediations:
+- name: "responsetime_p90_sockshop_carts"
+  actions:
+  - action: "featuretoggle"
+    value: "enablecache:on"
+```
+
+
+*   `name`: it is the name of the violated SLO or type of problem.
+*   `actions`: a list of actions that can be executed to remediate the problem
+*   `action`: name of the action as well as the value which should be set. I.e., the action should be set to `featuretoggle`, the value defines which feature toggle as well as the new value the feature toggle should be set to. In this case, the name of the feature toggle is `enablecache` and the value will be set to `on` once this remediation is executed.
+
+
+# User Walkthrough
+
+
+## Setup & Configure Keptn 
+
+As you would normally do with Keptn, create a project, onboard a service, and deploy your service into your Kubernetes cluster.
+
+<details><summary>Instructions</summary>
```

Comment:
What's with the formatting here?

Reply:
just some html to be able to collapse some instructions since they are not the core of the use case.